assets/prototype/domain/eventEditor/containers/mutations/useEntityMutation.js

```diff
@@ -0,0 +1,146 @@
+
+import { mutations } from './';
+import useMutators from './useMutators';
+
+/**
+ * @param {string} type Entity type name
+ */
+const useEntityMutation = (type, id = '') => {
+	const client = useApolloClient();
+	const [result, setResult] = useState({ called: false, loading: false });
+	const mutators = useMutators();
+
+	/**
+	 * @param {string} mutation Type of mutation - CREATE|UPDATE|DELETE
+	 */
+	const getMutation = (mutationType = 'UPDATE') => {
+		// For example "CREATE_DATETIME"
+		const mutaton = `${mutationType}_${type.toUpperCase()}`;
+		return mutations[mutaton];
+	};
+
+	/**
+	 * @param {string} mutation Type of mutation
+	 * @param {object} data     Mutation data
+	 */
+	const getMutationOptions = (mutationType, data = {}) => {
+		// e.g. "datetmeMutator"
+		const key = `${type.toLowerCase()}Mutator`;
+		const { [key]: mutator } = mutators;
+		const { variables, optimisticResponse, update } = mutator(mutationType, data);
+
+		return { variables, optimisticResponse, update };
+	};
+
+	/**
+	 * @param {object} data
+	 */
+	const getCreateMutation = (data) => {
+		const mutationType = 'CREATE';
+		const mutation = getMutation(mutationType);
+		const options = getMutationOptions(mutationType, data);
+
+		return { mutation, ...options };
+	};
+
+	/**
+	 * @param {string} id Entity id
+	 */
+	const getUpdateMutation = (data) => {
+		const mutationType = 'UPDATE';
+		const mutation = getMutation(mutationType);
+		const options = getMutationOptions(mutationType, { ...data, id });
+
+		return { mutation, ...options };
+	};
+
+	/**
+	 *
+	 */
+	const getDeleteMutation = () => {
+		const mutationType = 'DELETE';
+		const mutation = getMutation(mutationType);
+		const options = getMutationOptions(mutationType, { id });
+
+		return { mutation, ...options };
+	};
+
+	/**
+	 *
+	 */
+	const onMutationStart = () => {
+		setResult({
+			loading: true,
+			error: undefined,
+			data: undefined,
+			called: true,
+		});
+	};
+
+	/**
+	 *
+	 */
+	const onMutationComplete = (response, onCompleted) => {
+		const { data, errors } = response;
+		const error = errors && errors.length > 0 ? new ApolloError({ graphQLErrors: errors }) : undefined;
+
+		setResult({
+			called: true,
+			loading: false,
+			data,
+			error,
+		});
+
+		if (typeof onCompleted === 'function') {
+			onCompleted(data);
+		}
+	};
+
+	/**
+	 *
+	 */
+	const onMutationError = (error, onError) => {
+		setResult({
+			loading: false,
+			error,
+			data: undefined,
+			called: true,
+		});
+
+		if (typeof onError === 'function') {
+			onError(error);
+		}
+	};
+
+	/**
+	 *
+	 */
+	const mutate = (options) => {
+		const { onCompleted, onError, ...mutationOptions } = options;
+		onMutationStart();
+		client
+			.mutate(mutationOptions)
+			.then((response) => {
+				onMutationComplete(response, onCompleted);
+				return response;
+			})
+			.catch((error) => {
+				onMutationError(error, onError);
+				throw error;
+			});
+
+		return result;
```

Comment:
What do you think about:
```suggestion
  return client
			.mutate(mutationOptions)
			.then((response) => {
				onMutationComplete(response, onCompleted);
				return result;
			})
			.catch((error) => {
				onMutationError(error, onError);
				throw error;
			});
```

Reply:
`result` is not from there. It comes from the outer scope from `useState`

Reply:
where is `result` even being declared?
is this some JS magic I am yet to learn about?

Reply:
It's declared at the top:
```js
const [result, setResult] = useState({ called: false, loading: false });
```
Since its value is set by `setResult`, the consumer of `mutate` function will have realtime update of `loading`, `data`, `error` etc.